### DISCUSSION TITLE
Recipe 2 is up - demos new column (label) creation by Stat mechanism.

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [EvaMaeRey/easy-geom-recipes#10] [Show and tell]
@yjunechoe, I put recipe 2 up using webr chunks and callouts like recipe 1.  https://evamaerey.github.io/easy-geom-recipes/recipe2coordinates.html

I think there's enough going on that introducing default_aes and after_stat doesn't make sense.   Recipe one is about summarizing existing vars.  Recipe 2 is about adding vars in your compute (that happen to be natively understood by ggplot2).  I think this alone is kind of potentially illuminating without adding more to think about.  So the example is just creating dedicated layers that do their dedicated thing (add coords, add indices)

With the indexing case, you have a situation where you don't need required_aes, so that's something else to chew on.   I think the stat_index also teases compute by panel interest. 

I think going a different dircetion with these in a future recipe could be cool.   Label aes default really useful for lot of Stats, like packed circles and waterfall, so very generic xy annotation Stat could be a nice minimalist way of introducing the concept.

**Comment by yjunechoe** [comment]
Sorry I'm late to respond here!

> I think there's enough going on that introducing default_aes and after_stat doesn't make sense. Recipe one is about summarizing existing vars. Recipe 2 is about adding vars in your compute (that happen to be natively understood by ggplot2). I think this alone is kind of potentially illuminating without adding more to think about. So the example is just creating dedicated layers that do their dedicated thing (add coords, add indices)

Reasonable scope!

> I think going a different dircetion with these in a future recipe could be cool. Label aes default really useful for lot of Stats, like packed circles and waterfall, so very generic xy annotation Stat could be a nice minimalist way of introducing the concept.

This makes sense! In fact, on the note of "future recipe" I feel like you could bring the same recipe case-studies back for a deeper treatment of design considerations, like how/when to use default_aes, how to write ggproto methods using base-R and `{vctrs}` (vs. `{dplyr}`), the flexibility of using `after_stat()` and friends, etc. So you could imagine a structure like:

```
The Basics
├── Recipe 1: `geom_medians()` and `geom_means()`
├── Recipe 2: `geom_index()` and `geom_coordinates()`
└── Recipe 3: ...
Design Principles
├── Revisiting recipe 1 with `{vctrs}`
├── Revisiting recipe 2 with `after_stat()`
└── Revisiting recipe 3 with ...
```

Where "The basics" are available for an early exit (users who want to get stuff done), and "Design principles" are available for those who are interested in "best practices" (e.g., otherwise experienced R devs wanting to maintain an internal ggplot extension pkg). Just a thought - something for the future

---

On Recipe 2, specifically, it looks great - only a few minor suggestions from me - from top-to-bottom:

> that we don’t use “index” as the stat argument, which would be more consistent with base ggplot2 documentation. However, if you prefer, you can refer to your newly created Stat this way when testing, i.e. geom_point(stat = "index", size = 7)

It should be specified here that the ability to refer to the stat as `"index"` comes from the CamelCase convention (so if you break the naming convention, you no longer get the `StatIndex` <-> `"index"` equivalence)

> that StatIndex is used to define the new layer function, so the computation that defines it, which is to summarize to medians, will be in play before the layer is rendered.

"summarize to medians" seems to be a copy-paste that needs updating - a quick Ctrl+F of "medians" also show other places that need an update. Also this is opinionated, but I'd be a little careful with the term "render"/"rendering", as it means something very specific in graphics - I'd maybe suggest something more neutral, like "will be used by the layer".

> Final touch, list2 will error without export from rlang, so update to rlang::list2.

In the interests of least friction / early exit, I think you should also give users an option to use `list()` instead of `rlang::list2()`. They're equivalent sans some niche extra features in the rlang version, and you can simply point readers to the `list2()` docs to decide for themselves whether they want that extra overhead/dependency (I assume a typical newcomer to ggplot internals will not be familiar with rlang metaprogramming)

> You can also define geom with identical properties via aliasing.

I think the recipes would benefit from establishing some terminology early on. When referring to a `geom_*()` function (vs. a `Geom*` ggproto), you can say "constructor function" - this is consistent with terminology used in ggplot - https://ggplot2-book.org/ext-springs.html#constructors

> Pro-tip: ‘Express’ alternative for user-facing functions … (Caution, needs vetting)

If you don't get around to polishing it up to fit nicely with the flow, I think this callout can be omitted for this recipe (since there's already a lot here!). Might be worth reserving for the "design principles" kind of recipes in the future

> Your Turn: write geom_coordinates()

I might have missed it, but is there "solutions" anywhere? You can maybe hide them in a callout at the bottom of the section

**Reply by EvaMaeRey** [reply]
More later (so much here!). So, yes, sometimes I think about dplyr and the recipes!    I think it makes a lot of sense to use dplyr in introduction phase because of it's arguable success at reducing cognitive load (and also to have things be pretty parallel to 'step 0' - how you'd get things done without extension), so you can kind of focus on the new gg internals ideas.  And discussion of re-writes to avoid dependancies for packaging scenario sounds appropriate.  Thanks for this point!

**Reply by EvaMaeRey** [reply]
And to be totally honest, I don't know what a rewrite of any of my stuff would look like with vctrs, so if you can point me to an easy example, or whip up a micro example, I'm interested!

**Reply by yjunechoe** [reply]
> And to be totally honest, I don't know what a rewrite of any of my stuff would look like with vctrs, so if you can point me to an easy example, or whip up a micro example, I'm interested!

Me also to be honest - I don't know too well either! For the most part I think it's just simple, almost style-guide-like replacements like `vec_c()` instead of `c` and `data_frame0()` instead of `data.frame()`.

So maybe you don't have to talk about vctrs, but I think it'd be useful to have _some_ discussion about dependencies (dplyr is nice but you may not want to import that whole thing!) and nudging pkg developers to maybe consider low/no-dep code in vctrs or base-R (and show an example to that point, e.g. `aggregate()` vs. `summarize()`).

**Reply by EvaMaeRey** [reply]
From my perspective, it sounds like ggtedious (ggplot2 extension packaging best practices) workshop material at this point - and then maybe formalization happens after?

**Reply by yjunechoe** [reply]
Yeah and I don't even think this has to be such a large discussion for a "recipes" cookbook - just making readers aware of the fact that design principles for ggproto methods _are a thing_, so that they don't go trying to re-invent the wheel themselves, should be sufficient!

**Reply by EvaMaeRey** [reply]
Thanks @yjunechoe 

Regarding ggproto method design principles, I guess that a go-to resource isn't available (yet!)?

I do think the scope of 'Easy recipes' is introduction to fundamentals in a practical setting - so 1) create some real, usable functionality and make it feel logical 2) gain awareness/vocabulary of some Stat basics...  ggproto, Stat, Geom, compute_group, requried_aes.  All this is potentially foreign coming in - non-trivial footholds for further internals dives.

I think experimental packages with internal syntax like dplyr pipelines are not inherently problematic.  They can be updated to a more robust form if they see a lot of use?    

Personally, too, I am really interested in the ad-hoc, in script extension and ggproto use, and there, if you are using dplyr anyway, it makes total sense to use tidyverse w/ no qualms?

**Reply by yjunechoe** [reply]
Yeah you're right - that's fair! Maybe vctrs/base-R stuff can be saved for another time and place :)

**Reply by EvaMaeRey** [reply]
I want do/know design principles, though!  As far as codifying them, that'd be well down the road thing, and maybe separate project  from 'easy recipes'.   Would you say the design principles are mostly learned from reading ggplot2 repo or elsewhere discussed?

**Reply by yjunechoe** [reply]
>  Would you say the design principles are mostly learned from reading ggplot2 repo or elsewhere discussed?

I think mostly from reading code in additional to some general materials like https://design.tidyverse.org/

> that'd be well down the road thing, and maybe separate project from 'easy recipes'. 

Yeah I think you've made a good case for it - dependencies and type stability and such are low-level concerns that people can live without :)
